[#58426] add feature spec for custom fields of type hierarchy

Kharonus opened the pull request:
# Ticket
[OP#58426](https://community.openproject.org/work_packages/58426)
also fixes: [OP#58466](https://community.openproject.org/work_packages/58466)

# What are you trying to accomplish?
- UI and main use cases are covered by a feature spec

# What approach did you choose and why?
- add a couple of test selectors
- add page objects

Kharonus committed:
[#58426] add feature spec for custom fields of type hierarchy

- https://community.openproject.org/work_packages/58426
- add a couple of test selectors
- add page objects

mereghost commented on spec/features/custom_fields/hierarchy_custom_field_spec.rb:
```diff
@@ -0,0 +1,131 @@
+#++
+
+require "spec_helper"
+require "support/pages/custom_fields/hierarchy_page"
```
🟢 The support folder is already required by `spec_helper`, this shouldn't be needed.

Kharonus replied:
done

mereghost commented on app/components/admin/custom_fields/hierarchy/delete_item_dialog_component.html.erb:
```diff
@@ -28,7 +28,9 @@ See COPYRIGHT and LICENSE files for more details.
-  render(Primer::Alpha::Dialog.new(id: DIALOG_ID, title: "Delete item")) do |dialog|
+  render(Primer::Alpha::Dialog.new(id: DIALOG_ID,
+                                   title: "Delete item",
+                                   data: { test_selector: TEST_SELECTOR })) do |dialog|
```
❓ Does this end up on the production rendered HTML?

Kharonus replied:
I need to relay the question, as I remember I asked the same thing some time ago :D

Kharonus replied:
When looking at community, I still see the html attributes `data-test-selector`. So I guess, the answer is yes, it is part of production rendered HTML.

Does this lead to another issue, you had in mind?

mereghost replied:
Nope. I'm OK with it existing in production, I was just wondering if it got removed.

Some testing stuff is bound to escape, like the oil level dipstick on car engines. =)

mereghost commented on app/components/admin/custom_fields/hierarchy/items_component.html.erb:
```diff
@@ -32,7 +32,10 @@ See COPYRIGHT and LICENSE files for more details.
-          render Primer::Beta::Blankslate.new(border: true) do |component|
+          render Primer::Beta::Blankslate.new(
+            border: true,
+            test_selector: "op-custom-fields--hierarchy-items-blankslate"
```
🟡 Any reason this isn't a constant like the other one?

Kharonus replied:
Honestly there is no good reason to extract those inline selectors to variables. I did it for the dialog, because it was already in the template I copied.

But those selectors are unique. They are not reused.

mereghost commented on app/components/admin/custom_fields/hierarchy/new_item_form_component.html.erb:
```diff
@@ -31,6 +31,7 @@ See COPYRIGHT and LICENSE files for more details.
   primer_form_with(
     url: custom_field_items_path(@custom_field),
     method: :post,
+    data: { test_selector: "op-custom-fields--new-item-form" }
```
🟡 Same question as above

Kharonus replied:
see above

mereghost commented on app/components/custom_fields/details_component.html.erb:
```diff
@@ -1,9 +1,14 @@
+          render Primer::Alpha::Banner.new(scheme: :default,
+                                           icon: :info,
+                                           dismiss_scheme: :hide,
+                                           test_selector: "op-custom-fields--new-hierarchy-banner") do
```
🟡 Same same.

Kharonus replied:
see above

mereghost commented on app/components/custom_fields/details_component.rb:
```diff
@@ -32,5 +32,13 @@ module CustomFields
+
+    alias_method :custom_field, :model
+
+    def has_no_items_or_projects?
```
🟡 Can we leverage the `render?` method

Kharonus replied:
I'd say no, as I'd need to create a separate component just for the banner.

mereghost commented on app/views/custom_fields/_form.html.erb:
```diff
@@ -37,7 +37,10 @@ See COPYRIGHT and LICENSE files for more details.
+    <%= f.text_field :name,
+                     required: true,
+                     container_class: "-middle",
+                     "data-test-selector": "op-custom-fields--new-custom-field-name" %>
```
🟡 More non-constant selectors.

Kharonus replied:
see above

mereghost commented on spec/features/custom_fields/hierarchy_custom_field_spec.rb:
```diff
@@ -0,0 +1,131 @@
+  let(:new_custom_field_page) { Pages::CustomFields::NewPage.new }
+  let(:hierarchy_page) { Pages::CustomFields::HierarchyPage.new }
+
+  it "lets you create, update and delete a custom field of type hierarchy",
```
🔴 Wouldn't be better to split those cases into separate tests?

Kharonus replied:
No, IMHO that's exactly the opposite of what we should do with e2e tests.

Reasoning:
The ramp up of the test suite takes ages, compared to the actual test run. This is why I wrote the work with the custom field like a story. You start by creating it, do a lot of nasty stuff to it, and then you end by deleting it. Also, an expensive test like this is meant to test UI and integration, not all of the use cases. Which is why I kept some things out for now.

mereghost replied:
I'd separate them to avoid eventual state leaks to change the behaviour. 

I tend to see feature specs as user stories with clear intent in each test and acceptance criteria for it. We may need to involve others in this discussion.

@apfohl @brunopagno WDYT?

brunopagno replied:
In the past I was bitten by having a single/few tests because not all errors would be visible right away. Then you get to the situation where you fix something and it explodes again further into the test.

I understand the argument against, though, that tests might take much longer.

I guess it's a question if we can "pay the price" of longer running tests.

apfohl replied:
I see also two sides. The one side don't wanting an even longer running test suite and the other side wanting to have easier maintainable specs. It seems like you can't have both at the moment. Honestly I would say we should go the way of doing it like @Kharonus. Define the domain story (you could even draw it) and play that thing through. Maybe we could use `# region REGION_NAME` `# endregion` to visually highlight the steps. The IDE allows to collapse them.

In general I would not delay on feature spec development to much. Invest more into automated domain model unit testing.

Kharonus replied:
Is separating the test into clearly distinct `regions` something you would approve @mereghost ?

mereghost replied:
I'd say it would alleviate the issue at least. 👍🏼

mereghost commented on spec/support/pages/custom_fields/hierarchy_page.rb:
```diff
@@ -0,0 +1,117 @@
+        end
+      end
+
+      def expect_empty_items_banner(visible:)
```
😕 I have mixed feeling about hidden assertions in helpers as makes the tests kinda opaque. That being said, it is just me so don't bother too much with it.

Kharonus replied:
My goal was to make the test itself easily readable without bothering to much about capybara syntax. Things like `click_on` and `fills_in` are already doing that quite nicely, but the stuff about test_selectors add nothing to the readability imho.

mereghost replied:
`click_on` and `fill_in` represent user actions.

This sets an expectation/assertion making it completely invisible to the test as you scan for `expect(thing).to eq(zeug)` making parsing what is being tested pretty hard. =/

Kharonus replied:
Is it then a general opposition of using POs? They are meant to encapsulate these kind of things into more readable things.

mereghost replied:
Nah, I'm cool with PO's... even with managers. :P

Seriously, POs are ok the issue is the hidden assertion. When parsing specs, I search for `expect(something).to something` and not `page.expect_something`.

I classify this as worse than shared specs used only in one place.

Kharonus replied:
okay, understood. I'll recheck the code with having this kind of glasses on.

apfohl replied:
Maybe we can extend the `expect` language with new custom matchers. Not sure if that is easily doable. I'm thinking of some kind of fluent syntax extension:

```ruby
expect(page).when(:visible).to have(BANNER).when_not(:visible).not_to have(BANNER)
```

Or different.

mereghost reviewed (changes requested):
Bunch of questions, rants and of course, no suggestions. 🤣

Kharonus committed:
[#58426] add more structure to feature spec

- add regions to different steps of the feature spec
- removed expectations from page object
